File: kotlin/sdk/src/commonTest/kotlin/com/basecamp/sdk/services/CommentsMentionsTest.kt
package com.basecamp.sdk.services

import com.basecamp.sdk.BasecampClient
import com.basecamp.sdk.BasecampException
import com.basecamp.sdk.generated.comments
import com.basecamp.sdk.testBasecampClient
import io.ktor.client.engine.mock.MockEngine
import io.ktor.client.engine.mock.respond
import io.ktor.client.request.HttpRequestData
import io.ktor.content.TextContent
import io.ktor.http.ContentType
import io.ktor.http.HttpHeaders
import io.ktor.http.HttpStatusCode
import io.ktor.http.headersOf
import kotlinx.coroutines.test.runTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertTrue

class CommentsMentionsTest {

    private companion object {
        /** An attachable Person sgid for 1049715915, as BC3 serves it. */
        const val ANNIE_SGID =
            "BAh7CEkiCGdpZAY6BkVUSSIrZ2lkOi8vYmMzL1BlcnNvbi8xMDQ5NzE1OTE1P2V4cGlyZXNfaW4GOwBUSSIMcHVycG9zZQY7AFRJ" +
                "Ig9hdHRhY2hhYmxlBjsAVEkiD2V4cGlyZXNfYXQGOwBUMA==--919d2c8b11ff403eefcab9db42dd26846d0c3102"

        /** And one for 1049715914. */
        const val VICTOR_SGID =
            "BAh7CEkiCGdpZAY6BkVUSSIrZ2lkOi8vYmMzL1BlcnNvbi8xMDQ5NzE1OTE0P2V4cGlyZXNfaW4GOwBUSSIMcHVycG9zZQY7AFRJ" +
                "Ig9hdHRhY2hhYmxlBjsAVEkiD2V4cGlyZXNfYXQGOwBUMA==--aabbccdd"
    }

    private val paths = mutableListOf<String>()
    private val bodies = mutableListOf<String?>()

    private fun personJson(id: Long, sgid: String) =
        """{"id": $id, "name": "Person $id", "attachable_sgid": "$sgid"}"""

    private fun commentJson(content: String) = """{
        "id": 1069479370, "status": "active", "visible_to_clients": false,
        "created_at": "2022-10-31T14:22:33.169Z", "updated_at": "2022-10-31T14:22:33.169Z",
        "title": "Re: We won Leto!", "inherits_status": true, "type": "Comment",
        "url": "https://3.basecampapi.com/999/buckets/1/comments/1069479370.json",
        "app_url": "https://3.basecamp.com/999/buckets/1/comments/1069479370",
        "parent": {"id": 1, "title": "We won Leto!", "type": "Message",
                   "url": "https://3.basecampapi.com/999/buckets/1/messages/1.json",
                   "app_url": "https://3.basecamp.com/999/buckets/1/messages/1"},
        "bucket": {"id": 1, "name": "The Leto Laptop", "type": "Project"},
        "creator": {"id": 1049715915, "name": "Annie Bryan"},
        "content": "$content",
        "content_attachments": []
    }"""

    private fun client(route: (HttpRequestData) -> Pair<HttpStatusCode, String>): BasecampClient {
        val engine = MockEngine { request ->
            paths.add(request.url.encodedPath)
            bodies.add((request.body as? TextContent)?.text)
            val (status, body) = route(request)
            respond(
                content = body,
                status = status,
                headers = headersOf(HttpHeaders.ContentType, ContentType.Application.Json.toString()),
            )
        }
        return testBasecampClient {
            accessToken("test-token")
            this.engine = engine
        }
    }

    @Test
    fun resolvesEveryPersonBeforePostingAndWritesTheMentionFromAttachableSgid() = runTest {
        val client = client { request ->
            if (request.url.encodedPath.startsWith("/999/people/")) {
                HttpStatusCode.OK to personJson(1049715915, ANNIE_SGID)
            } else {
                HttpStatusCode.Created to commentJson("posted")
            }
        }
        client.forAccount("999").comments.createWithMentions(1, "<div>On it.</div>", listOf(1049715915))
        assertEquals(listOf("/999/people/1049715915", "/999/recordings/1/comments.json"), paths)
        assertTrue(
            bodies.last().orEmpty().contains("<bc-attachment sgid=\\\"$ANNIE_SGID\\\"></bc-attachment> On it."),
            "the mention is placed inside the content's first block: ${bodies.last()}",
        )
        client.close()
    }

    @Test
    fun readsEachDistinctIdOnceAndKeepsTheRequestedOrder() = runTest {
        val client = client { request ->
            val path = request.url.encodedPath
            when {
                path.endsWith("/1049715915") -> HttpStatusCode.OK to personJson(1049715915, ANNIE_SGID)
                path.endsWith("/1049715914") -> HttpStatusCode.OK to personJson(1049715914, VICTOR_SGID)
                else -> HttpStatusCode.Created to commentJson("posted")
            }
        }
        val expanded = client.forAccount("999").comments.expandMentions(
            "Hi",
            listOf(1049715915, 1049715914, 1049715915),
        )
        assertEquals(listOf("/999/people/1049715915", "/999/people/1049715914"), paths)
        assertEquals(
            "<bc-attachment sgid=\"$ANNIE_SGID\"></bc-attachment> " +
                "<bc-attachment sgid=\"$VICTOR_SGID\"></bc-attachment> Hi",
            expanded,
        )
        client.close()
    }

    @Test
    fun readsThePersonEvenWhenTheContentAlreadyCarriesAMatchingTag() = runTest {
        // The trust boundary: an sgid in caller-supplied content is unsigned, so
        // it can never stand in for the authoritative people read. The read still
        // happens; withMentions then adds nothing, because the sgid the read
        // returned is the exact string already present.
        val client = client { HttpStatusCode.OK to personJson(1049715915, ANNIE_SGID) }
        val content = "<div><bc-attachment sgid=\"$ANNIE_SGID\"></bc-attachment> already</div>"
        val expanded = client.forAccount("999").comments.expandMentions(content, listOf(1049715915))
        assertEquals(listOf("/999/people/1049715915"), paths, "the people read is never skipped")
        assertEquals(content, expanded)
        client.close()
    }

    @Test
    fun aFailedPersonReadPostsNothing() = runTest {
        val client = client { HttpStatusCode.NotFound to """{"error":"Not found"}""" }
        assertFailsWith<BasecampException.NotFound> {
            client.forAccount("999").comments.createWithMentions(1, "<div>Hi</div>", listOf(1049715915))
        }
        assertEquals(listOf("/999/people/1049715915"), paths, "nothing is posted on a partial mention list")
        client.close()
    }

    @Test
    fun refusesAnInvalidMentionIdAndEmptyContentBeforeAnyRequest() = runTest {
        val client = client { error("no request expected") }
        val service = client.forAccount("999").comments
        assertFailsWith<BasecampException.Usage> { service.expandMentions("Hi", listOf(0)) }
        assertFailsWith<BasecampException.Usage> { service.createWithMentions(1, "", listOf(1049715915)) }
        assertTrue(paths.isEmpty())
        client.close()
    }

    @Test
    fun noMentionsMeansNoPeopleReads() = runTest {
        val client = client { HttpStatusCode.Created to commentJson("posted") }
        client.forAccount("999").comments.createWithMentions(1, "<div>Hi</div>", emptyList())
        assertEquals(listOf("/999/recordings/1/comments.json"), paths)
        client.close()
    }
}

File: typescript/tests/services/comments-mentions.test.ts
/**
 * Tests for the mention composites on CommentsService
 * (src/services/comments-extensions.ts).
 *
 * The conformance fixture pins the happy path — one people read, then the
 * write, with the tag placed inside the content's first block. These carry the
 * rest: the read is never skipped on the strength of an unsigned sgid already
 * in the content, a failed lookup posts nothing, and a repeated id costs one
 * read.
 */
import { describe, it, expect, beforeEach } from "vitest";
import { http, HttpResponse } from "msw";
import { server } from "../setup.js";
import { createBasecampClient } from "../../src/client.js";
import type { BasecampClient } from "../../src/client.js";
import { BasecampError } from "../../src/errors.js";
import { CommentsService } from "../../src/index.js";
import { personSGID } from "../helpers/sgid.js";

const BASE_URL = "https://3.basecampapi.com/12345";
const VICTOR = 1049715914;
const ANNIE = 1049715915;

const VICTOR_SGID = personSGID(VICTOR);
const ANNIE_SGID = personSGID(ANNIE);

const attachment = (sgid: string): string => `<bc-attachment sgid="${sgid}"></bc-attachment>`;

const personBody = (id: number, sgid: string) => ({
  id,
  attachable_sgid: sgid,
  name: `Person ${id}`,
  email_address: `p${id}@honchodesign.com`,
  personable_type: "User",
});

const commentBody = (content: string) => ({
  id: 1069479370,
  status: "active",
  visible_to_clients: false,
  created_at: "2022-10-31T14:22:33.169Z",
  updated_at: "2022-10-31T14:22:33.169Z",
  title: "Re: We won Leto!",
  inherits_status: true,
  type: "Comment",
  url: `${BASE_URL}/comments/1069479370.json`,
  app_url: "https://3.basecamp.com/12345/buckets/1/comments/1069479370",
  parent: { id: 1, title: "We won Leto!", type: "Message", url: "", app_url: "" },
  bucket: { id: 1, name: "The Leto Laptop", type: "Project" },
  creator: personBody(VICTOR, VICTOR_SGID),
  content,
  content_attachments: [],
});

/**
 * Records every (method, path) MSW served, in order.
 *
 * Synchronous on purpose: a listener that awaited the request body would push
 * its entry after the assertion ran. Written bodies are captured in the POST
 * handler itself, where the await is part of serving the request.
 */
function trackRequests(): { calls: string[] } {
  const calls: string[] = [];
  server.events.removeAllListeners("request:start");
  server.events.on("request:start", ({ request }) => {
    calls.push(`${request.method} ${new URL(request.url).pathname}`);
  });
  return { calls };
}

describe("comments mention composites", () => {
  let client: BasecampClient;

  beforeEach(() => {
    client = createBasecampClient({
      accountId: "12345",
      accessToken: "test-token",
      enableRetry: false,
    });
  });

  describe("expandMentions", () => {
    it("resolves each id and places the tag inside the first block", async () => {
      const tracked = trackRequests();
      server.use(http.get(`${BASE_URL}/people/${VICTOR}`, () => HttpResponse.json(personBody(VICTOR, VICTOR_SGID))));

      const expanded = await client.comments.expandMentions("<div>On it.</div>", [VICTOR]);

      expect(expanded).toBe(`<div>${attachment(VICTOR_SGID)} On it.</div>`);
      expect(tracked.calls).toEqual([`GET /12345/people/${VICTOR}`]);
    });

    it("reads each distinct id once, in order", async () => {
      const tracked = trackRequests();
      server.use(
        http.get(`${BASE_URL}/people/${VICTOR}`, () => HttpResponse.json(personBody(VICTOR, VICTOR_SGID))),
        http.get(`${BASE_URL}/people/${ANNIE}`, () => HttpResponse.json(personBody(ANNIE, ANNIE_SGID))),
      );

      const expanded = await client.comments.expandMentions("<div>Hi</div>", [VICTOR, ANNIE, VICTOR]);

      expect(tracked.calls).toEqual([`GET /12345/people/${VICTOR}`, `GET /12345/people/${ANNIE}`]);
      expect(expanded).toBe(`<div>${attachment(VICTOR_SGID)} ${attachment(ANNIE_SGID)} Hi</div>`);
    });

    it("reads the person even when the content already names them, and keeps the read's sgid", async () => {
      // The trust boundary: an sgid found in caller-supplied content is
      // unsigned, so it can neither prove the mention exists nor stand in for
      // the people read. A stale tag naming the right person is left where it
      // is, and the authoritative mention is added alongside it.
      const stale = personSGID(VICTOR, { digest: "f".repeat(40) });
      expect(stale).not.toBe(VICTOR_SGID);
      const tracked = trackRequests();
      server.use(http.get(`${BASE_URL}/people/${VICTOR}`, () => HttpResponse.json(personBody(VICTOR, VICTOR_SGID))));

      const expanded = await client.comments.expandMentions(`<div>${attachment(stale)} On it.</div>`, [VICTOR]);

      expect(tracked.calls).toEqual([`GET /12345/people/${VICTOR}`]);
      expect(expanded).toBe(`<div>${attachment(VICTOR_SGID)} ${attachment(stale)} On it.</div>`);
    });

    it("adds nothing when the content already carries the exact sgid the read returned", async () => {
      server.use(http.get(`${BASE_URL}/people/${VICTOR}`, () => HttpResponse.json(personBody(VICTOR, VICTOR_SGID))));
      const content = `<div>${attachment(VICTOR_SGID)} On it.</div>`;
      expect(await client.comments.expandMentions(content, [VICTOR])).toBe(content);
    });

    it("makes no request when no one is mentioned", async () => {
      const tracked = trackRequests();
      expect(await client.comments.expandMentions("<div>On it.</div>", [])).toBe("<div>On it.</div>");
      expect(tracked.calls).toEqual([]);
    });

    it("refuses a non-positive id before any request", async () => {
      const tracked = trackRequests();
      await expect(client.comments.expandMentions("<div>Hi</div>", [0])).rejects.toThrow(/invalid mention person id/);
      await expect(client.comments.expandMentions("<div>Hi</div>", [-3])).rejects.toThrow(BasecampError);
      expect(tracked.calls).toEqual([]);
    });

    it("fails the expansion when a person read fails, naming the mention it was resolving", async () => {
      server.use(
        http.get(`${BASE_URL}/people/${VICTOR}`, () => HttpResponse.json({ error: "Not found" }, { status: 404 })),
      );

      const err = await client.comments
        .expandMentions("<div>Hi</div>", [VICTOR])
        .catch((e: unknown) => e);

      expect(err).toBeInstanceOf(BasecampError);
      // The read's classification survives the added context.
      expect((err as BasecampError).code).toBe("not_found");
      expect((err as BasecampError).message).toContain(`resolving mention for person ${VICTOR}`);
    });

    it("fails when a person carries no attachable_sgid to mention", async () => {
      server.use(http.get(`${BASE_URL}/people/${VICTOR}`, () => HttpResponse.json({ id: VICTOR, name: "Victor" })));
      await expect(client.comments.expandMentions("<div>Hi</div>", [VICTOR])).rejects.toThrow(/no attachable_sgid/);
    });

    it("explains itself when the service was built without the client's people read", async () => {
      const bare = new CommentsService(client.raw);
      await expect(bare.expandMentions("<div>Hi</div>", [VICTOR])).rejects.toThrow(/createBasecampClient/);
    });
  });

  describe("createWithMentions", () => {
    it("resolves every mention before posting, and writes the expanded content", async () => {
      const tracked = trackRequests();
      const posted: unknown[] = [];
      server.use(
        http.get(`${BASE_URL}/people/${VICTOR}`, () => HttpResponse.json(personBody(VICTOR, VICTOR_SGID))),
        http.post(`${BASE_URL}/recordings/1069479351/comments.json`, async ({ request }) => {
          posted.push(await request.json());
          return HttpResponse.json(commentBody(`<div>${attachment(VICTOR_SGID)} On it.</div>`), { status: 201 });
        }),
      );

      const comment = await client.comments.createWithMentions(1069479351, "<div>On it.</div>", [VICTOR]);

      expect(tracked.calls).toEqual([
        `GET /12345/people/${VICTOR}`,
        "POST /12345/recordings/1069479351/comments.json",
      ]);
      expect(posted).toEqual([{ content: `<div>${attachment(VICTOR_SGID)} On it.</div>` }]);
      expect(comment.id).toBe(1069479370);
    });

    it("refuses a content that is not a string, and posts nothing", async () => {
      // The guard reads the value that goes out, not a coerced copy of it. `{}`
      // and `[]` are truthy, so a plain-JavaScript caller could sail past a
      // truthiness check into the mention walk — a raw TypeError with mentions
      // requested, and with none, a POST body the reference cannot produce,
      // since Go's signature takes a string. The second assertion is the one
      // that matters: an exception-only test passes in the world where the
      // write went out anyway.
      const tracked = trackRequests();
      server.use(
        http.post(`${BASE_URL}/recordings/1069479351/comments.json`, () =>
          HttpResponse.json(commentBody("<div>posted</div>"), { status: 201 }),
        ),
      );

      for (const content of [{}, [], 42, true, null, undefined] as unknown[]) {
        const err = await client.comments
          .createWithMentions(1069479351, content as string, [])
          .catch((e: unknown) => e);
        expect(err, JSON.stringify(content ?? null)).toBeInstanceOf(BasecampError);
        expect((err as BasecampError).code).toBe("usage");
      }
      expect(tracked.calls.some((call) => call.startsWith("POST"))).toBe(false);
    });

    it("posts nothing when a mention lookup fails", async () => {
      const tracked = trackRequests();
      server.use(
        http.get(`${BASE_URL}/people/${VICTOR}`, () => HttpResponse.json(personBody(VICTOR, VICTOR_SGID))),
        http.get(`${BASE_URL}/people/${ANNIE}`, () => HttpResponse.json({ error: "Forbidden" }, { status: 403 })),
      );

      await expect(
        client.comments.createWithMentions(1069479351, "<div>On it.</div>", [VICTOR, ANNIE]),
      ).rejects.toThrow(BasecampError);

      expect(tracked.calls.some((call) => call.startsWith("POST"))).toBe(false);
    });

    it("posts the content unchanged when no one is mentioned", async () => {
      const tracked = trackRequests();
      const posted: unknown[] = [];
      server.use(
        http.post(`${BASE_URL}/recordings/1069479351/comments.json`, async ({ request }) => {
          posted.push(await request.json());
          return HttpResponse.json(commentBody("<div>On it.</div>"), { status: 201 });
        }),
      );

      await client.comments.createWithMentions(1069479351, "<div>On it.</div>");

      expect(tracked.calls).toEqual(["POST /12345/recordings/1069479351/comments.json"]);
      expect(posted).toEqual([{ content: "<div>On it.</div>" }]);
    });

    it("refuses empty content before any request", async () => {
      const tracked = trackRequests();
      await expect(client.comments.createWithMentions(1069479351, "", [VICTOR])).rejects.toThrow(
        /comment content is required/,
      );
      expect(tracked.calls).toEqual([]);
    });
  });
});
